src/ModelContextProtocol/Protocol/Messages/ProgressToken.cs

```diff
@@ -0,0 +1,102 @@
+﻿using ModelContextProtocol.Utils;
+using System.ComponentModel;
+using System.Globalization;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace ModelContextProtocol.Protocol.Messages;
+
+/// <summary>
+/// Represents a progress token, which can be either a string or an integer.
+/// </summary>
+[JsonConverter(typeof(Converter))]
+public readonly struct ProgressToken : IEquatable<ProgressToken>
+{
+    /// <summary>The id, either a string or a boxed long or null.</summary>
+    private readonly object? _id;
+
+    /// <summary>Initializes a new instance of the <see cref="ProgressToken"/> with a specified value.</summary>
+    /// <param name="value">The required ID value.</param>
+    public ProgressToken(string value)
+    {
+        Throw.IfNull(value);
```

Comment:
Should we consider normalizing quoted integers into integers?

Reply:
If we assume that equality should follow JSON semantics, then the answer is probably no.